core/src/main/java/org/infinispan/interceptors/impl/PrefetchInterceptor.java

```diff
@@ -194,50 +199,45 @@ protected Object handleReadCommand(InvocationContext ctx, DataCommand command) t
-      GetCacheEntryCommand cmd = (GetCacheEntryCommand) command1;
+   private InvocationStage lookupLocalAndRetrieveRemote(InvocationContext ctx, Object key, TopologyAffectedCommand cmd, int segment) {
+      // We need to lookup the dataContainer directly as GetCacheEntryCommand won't return entry with null value
+      InternalCacheEntry entry = dataContainer.get(segment, key);
       if (trace) {
-         log.tracef("Locally prefetched entry %s", rv);
+         log.tracef("Locally prefetched entry %s", entry);
       }
-      CacheEntry entry = (CacheEntry) rv;
       Metadata metadata = entry != null ? entry.getMetadata() : null;
       CompletionStage<InternalCacheValue> future;
-      int segment = cmd.getSegment();
       if (metadata != null && metadata.version() != null && svm.isVersionActual(segment, metadata.version())) {
+         entryFactory.wrapExternalEntry(ctx, key, entry, true, true);
          return null;
       } else if ((metadata instanceof RemoteMetadata) &&
             svm.getSegmentState(segment) == ScatteredVersionManager.SegmentState.VALUE_TRANSFER) {
          // The RemoteMetadata is valid only during value transfer - in blocked state there shouldn't be any such
          // entry and during key transfer we could see metadata pointing to a node with outdated information.
          Address backup = ((RemoteMetadata) metadata).getAddress();
-         future = retrieveRemoteValue(Collections.singleton(backup), cmd.getKey(), segment, cmd);
+         future = retrieveRemoteValue(Collections.singleton(backup), key, segment, cmd);
       } else {
-         future = retrieveRemoteValue(null, cmd.getKey(), segment, cmd);
+         future = retrieveRemoteValue(null, key, segment, cmd);
       }
       return asyncValue(future.thenAccept(maxValue -> {
          if (maxValue == null) {
             return;
          }
-         PutKeyValueCommand putKeyValueCommand =
-               commandsFactory.buildPutKeyValueCommand(cmd.getKey(), maxValue.getValue(), segment, maxValue.getMetadata(),
-                                                       STATE_TRANSFER_FLAGS);
+         // The put below could fail updating the context if the data container got the updated value while we were
+         // prefetching that (then the version would not be higher than the on in DC).
+         // We need to call RepeatableReadEntry.updatePreviousValue() (through wrapExternalEntry) to get return value
+         // from the main command correct.
+         entryFactory.wrapExternalEntry(ctx, key, maxValue.toInternalCacheEntry(key), true, true);
+         PutKeyValueCommand putKeyValueCommand = commandsFactory.buildPutKeyValueCommand(
+               key, maxValue.getValue(), segment, maxValue.getMetadata(), STATE_TRANSFER_FLAGS);
          putKeyValueCommand.setTopologyId(cmd.getTopologyId());
-         invokeNext(ctx1, putKeyValueCommand);
+         invokeNext(ctx, putKeyValueCommand);
```

Comment:
Oops, this is probably a bug - if the command blocks in this call we're going to return too early.

Reply:
I think we may have discussed this, but back then you thought it couldn't block :)

I think you can leave the "naked" `invokeNext` if you remove the `setTopologyId` line above, so that the command can't be sent remotely.